src/equations/compressible_navier_stokes_3d.jl

```diff
@@ -0,0 +1,380 @@
+
+    return SVector(g1, g2, g3, g4, g5)
+  else # if orientation == 3
+    # viscous flux components in the z-direction
+    # Note, symmetry is exploited for tau_13 = tau_31, tau_23 = tau_32
+    h1 = zero(rho)
+    h2 = tau_13 * mu # tau_31 * mu
+    h3 = tau_23 * mu # tau_32 * mu
+    h4 = tau_33 * mu
+    h5 = ( v1 * tau_13 + v2 * tau_23 + v3 * tau_33 + q3 ) * mu
+
+    return SVector(h1, h2, h3, h4, h5)
+  end
+end
+
+
+# Convert conservative variables to primitive
+@inline function cons2prim(u, equations::CompressibleNavierStokesDiffusion3D)
+  rho, rho_v1, rho_v2, rho_v3, _ = u
+
+  v1 = rho_v1 / rho
+  v2 = rho_v2 / rho
+  v3 = rho_v3 / rho
+  T  = temperature(u, equations)
+
+  return SVector(rho, v1, v2, v3, T)
+end
+
+# Convert conservative variables to entropy
+# TODO: parabolic. We can improve efficiency by not computing w_1, which involves logarithms
+# This can be done by specializing `cons2entropy` and `entropy2cons` to `CompressibleNavierStokesDiffusion2D`,
+# but this may be confusing to new users.
+cons2entropy(u, equations::CompressibleNavierStokesDiffusion3D) = cons2entropy(u, equations.equations_hyperbolic)
+entropy2cons(w, equations::CompressibleNavierStokesDiffusion3D) = entropy2cons(w, equations.equations_hyperbolic)
+
+# the `flux` function takes in transformed variables `u` which depend on the type of the gradient variables.
+# For CNS, it is simplest to formulate the viscous terms in primitive variables, so we transform the transformed
+# variables into primitive variables.
+@inline function convert_transformed_to_primitive(u_transformed, equations::CompressibleNavierStokesDiffusion3D{GradientVariablesPrimitive})
+  return u_transformed
+end
+
+# TODO: parabolic. Make this more efficient!
+@inline function convert_transformed_to_primitive(u_transformed, equations::CompressibleNavierStokesDiffusion3D{GradientVariablesEntropy})
+  # note: this uses CompressibleNavierStokesDiffusion3D versions of cons2prim and entropy2cons
+  return cons2prim(entropy2cons(u_transformed, equations), equations)
+end
+
+
+# Takes the solution values `u` and gradient of the entropy variables (w_2, w_3, w_4, w_5) and
+# reverse engineers the gradients to be terms of the primitive variables (v1, v2, v3, T).
+# Helpful because then the diffusive fluxes have the same form as on paper.
+# Note, the first component of `gradient_entropy_vars` contains gradient(rho) which is unused.
+# TODO: parabolic; entropy stable viscous terms
+@inline function convert_derivative_to_primitive(u, gradient, ::CompressibleNavierStokesDiffusion3D{GradientVariablesPrimitive})
+  return gradient
+end
+
+# the first argument is always the "transformed" variables.
+@inline function convert_derivative_to_primitive(w, gradient_entropy_vars,
+                                                 equations::CompressibleNavierStokesDiffusion3D{GradientVariablesEntropy})
+
+  # TODO: parabolic. This is inefficient to pass in transformed variables but then transform them back.
```

Comment:
Is the same issue present in 2D? If so can we open an issue for this?

Reply:
No, as far as I know this is still on the meta triage list for the parabolic terms implementation #1147

Reply:
I did not find it, but have added it. /r